### ISSUE TITLE
Determine if libdispatch supports thread pooling natively.

### ISSUE THREAD

**Issue by rolivieri** [IBM-Swift/Kitura#149]
We should determine if libdispatch can be configured to use a thread pool for processing of async tasks. As of now, we believe that for each new async task that is created, libdispatch creates a new thread and assigns to it the new async task. If instead, libdispatch had a pool of threads ready to handle the processing of async tasks, better performance could be obtained.

**Comment by rfdickerson** [comment]
Haven't fully discovered an answer yet, but:

> Dispatch Queues are objects that maintain a queue of tasks, either anonymous code blocks or functions, and execute these tasks in their turn. The library automatically creates several queues with different priority levels that execute several tasks concurrently, selecting the optimal number of tasks to run based on the operating environment. A client to the library may also create any number of serial queues, which execute tasks in the order they are submitted, one at a time.[12] Because a serial queue can only run one task at a time, each task submitted to the queue is critical with regard to the other tasks on the queue, and thus a serial queue can be used instead of a lock on a contended resource.

**Comment by hassila** [comment]
It does support pooling if built with libpwq - which I believe has been integrated recently as a git submodule. If using a non overcommit dispatch queue the thread pool will use heuristics to size itself (allowing for cheap creation up to a certain number of threads), then it will check the number of running threads by introspecting /prof and allow additional threads to be created up until the number of available cores on the machine. If threads are sleeping on e g condition variables or are blocked in system calls, it will create additional ones. If the thread pool size grows a lot due to many such sleeping threads, they will be dropped away with hysteresis (to avoid expensive ramp up if there is spikey loads). 

(I was involved in writing those parts of pthread work queue a few years ago)

**Comment by hassila** [comment]
Or to put it in another way: make sure you have dispatch built properly with pthread work queue to get the behavior you would expect.

**Comment by hassila** [comment]
https://github.com/apple/swift-corelibs-libdispatch/commit/be83e850ed602414df3007ab2f380e33a079243d

**Comment by hassila** [comment]
I haven't  tried with later source drops, but just a hint if you are testing this:

If you have:
"libpthread_workqueue.so.0.0`overcommit_worker_main+0xa1"

in your thread backtraces, you are getting overcommit queues (you might get this if creating queues with default attributes), which means you will get too many threads.

If this is the case, you may want to explicitly set the target global queue to whatever priority you think is relevant, e.g.:
 `dq = dispatch_queue_create(queue_name, NULL);
	dispatch_set_target_queue(dq, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0));`

To instead target a non-overcommit queue (this was at least required a few years ago).

**Comment by rfdickerson** [comment]
Yes! Thank you for the message. We end up getting overcommit workers queues quite often. I will check the priority settings and do some tests.

**Comment by rfdickerson** [comment]
I ran some experiments on GTD using htop. In my examples, I queued up a large number of asynchronous tasks inside of a custom queue:

1) If the tasks do not use the CPU (I simply used libc sleep) the number of simultaneously running and dispatched tasks seemed unbounded by the number of tasks in enqueued. 

2) If the tasks contained CPU instructions (I used several thousand floating-point division instructions), GTD dispatched between 8-12 threads at a time. 

In summary, GTD seems to make some intelligent decisions regarding the number of threads allowed to be dispatched based on CPU utilization. It seems to me that this is also what we want, since most of our tasks in Kitura will be blocking instructions waiting for IO on a file descriptor. There will be a large number of user-level threads spawned in a real-life web server scenario.
